README.md
# Landing Page Info Aggregator

[![Build Status](https://travis-ci.org/reportportal/landing-aggregator.svg?branch=master)](https://travis-ci.org/reportportal/landing-aggregator)
[![Go Report Card](https://goreportcard.com/badge/github.com/reportportal/landing-aggregator)](https://goreportcard.com/report/github.com/reportportal/landing-aggregator)
[![Docker Pulls](https://img.shields.io/docker/pulls/reportportal/landing-aggregator.svg?maxAge=159200)](https://hub.docker.com/r/reportportal/landing-aggregator/)

[![Run in Postman](https://run.pstmn.io/button.svg)](https://app.getpostman.com/run-collection/39ce87bf716162454c2e)

Micro-service which serves next:
* cache tweets from https://twitter.com/ReportPortal_io
* return latest version numbers for ReportPortal services
* aggregates ReportPortal's GitHub statistics

## Build

```
make help:
	@echo "vendor      - Install govendor and sync vendored dependencies"
	@echo "checkstyle  - executes bunch of checkstyle validators"
	@echo "fmt         - formats the project"
	@echo "test        - executes unit tests"
	@echo "build       - builds binary"
	@echo "docker      - Builds docker image"
	@echo "clean       - Cleans build-related files from working directory"
	@echo "release     - Builds docker container and pushes new version to DockerHUB"
```
```bash make docker ``` 
builds image with name 'reportportal/landing-info'
To create container execute 
 
```bash docker run --name landing-info -p 8080:8080 reportportal/landing-info```
 
## API
 
 Aggregator exposes two endpoints:
 
```/twitter```
Returns cache of tweets searched by provided in configuration hashtag

### Twitter aggregation details
Aggregator supports two modes:
* 'Hashtag' or streaming mode. Enabled by default. Buffers all tweets found by provided search term.
Uses Twitter Streaming API to keep buffer up to date
* 'Follow user' mode. Enabled by providing search term starting from '@'. Buffers all messages of specified user except retweets and replies.
Uses long-pooling to keep buffer up to date.
 
```/versions```
Returns latest versions of ReportPortal's Docker Images. Obtains this information from GitHUB API

```/```
Returns all the cached and aggregated data including tweets from Twitter and GitHub-related info

### Github aggregation details
```github/contribution```
Returns commits and unique contributors for the last weeks
For example,
```json
{
  "commits": {
    "1": 0,
    "4": 201,
    "12": 1018
  },
  "unique_contributors": {
    "1": 20,
    "4": 52,
    "12": 80
  }
}
```
means that there were no commits for the current week, 201 commits for the last 4 weeks, etc.


```github/stars```
Returns stars count for each repository and total count

```/github/issues```
Aggregates issue statistics from each organization repository

## Configuration
Aggregator can be configured through env variables. The following configuration options are available:

| ENV VAR                       | Default Value    | Required    | Description                  |
| ------------------------------|:----------------:| -----------:|-----------------------------:|
| PORT                          | 8080             | false       |Application port              |
| TWITTER_CONSUMER              |                  | true        |Twitter API consumer key      |
| TWITTER_CONSUMER_SECRET       |                  | true        |Twitter API consumer secret   |
| TWITTER_TOKEN                 |                  | true        |Twitter API token             |
| TWITTER_TOKEN_SECRET          |                  | true        |Twitter API token  secret|
| TWITTER_BUFFER_SIZE|10|false|Tweets buffer size|
| TWITTER_SEARCH_TERM|@reportportal_io|false|Tweets search term|
| GITHUB_INCLUDE_BETA|false|false|Whether BETA versions should be included|
| GITHUB_TOKEN||false|GitHUB API Token| 

## Production deployment
Several instances of app should be deployed to provide fault-tolerance and distribute load.
There is [Traefik](traefik.io) which is reverse-proxy and load-balancer which does not require any service registry and allows zero-conf 
discovery via Docker API. Example can be found [here](docker-compose.yml)


CHANGELOG.md
# Changelog

## 1.0
##### Released: 13 March 2017

* Initial DockerHub Release 

## 1.1
##### Released: 16 March 2017

### New Features

* 404/Not Found handler
* Low memory consumption by caching only needed tweet info
* add changelog

### Bugfixes

* build script: Release to DockerHUB prior to making git tag

## v1.5
##### Released: 31 March 2017

### New Features

* 'INCLUDE_BETA' env variable to configre whether BETA versions should be included in the list 
* Twitter data now contains entities and extended_entities

## v1.18
##### Released: 17 May 2017

### New Features

*  Replace streaming with long-polling for 'follow' mode

## v1.19
##### Released: 17 May 2017

### Bugfixes

*  Do not add count to the requests into 'follow' mode since twitter includes retweets in this case.
 See [the docs](https://dev.twitter.com/rest/reference/get/statuses/user_timeline)
 
## v1.20
##### Released: 17 May 2017

### Bugfixes

*  Traverse result in reverse order to make newest tweets at the bottom of the buffer 

## v1.21
##### Released: 17 May 2017

### New Features

*  Enable extended tweet mode
 
## v1.22
##### Released: 18 May 2017
### Bugixes
*  Fix presence of replies in 'follow' user mode 

## v1.24
##### Released: 23 May 2017
### New Features
*  Cache stars count for each org repository

## v1.27
##### Released: 29 May 2017
### New Features
* GitHub statistics aggregation
* Composite root endpoint for all aggregated data


## v1.28
##### Released: 8 June 2017
### Bugfixes
* Only 30 first organization repositories were processed by aggregator

## v1.31
##### Released: 10 June 2017
### Bugfixes
* Remove profiler

## v1.32
##### Released: 8 Aug 2017
### Bugfixes
* Add retweets to tweets cache

## v1.34
##### Released: 8 Aug 2017
### Bugfixes
* Improved dockerfile


landinginfo.go
package main

import (
	"github.com/caarlos0/env"
	"github.com/reportportal/commons-go/commons"
	"github.com/reportportal/landing-aggregator/info"
	"goji.io"
	"goji.io/pat"

	log "github.com/sirupsen/logrus"
	"net/http"

	"os"
	"strconv"
)

var (
	// Branch contains the current Git revision. Use make to build to make
	// sure this gets set.
	Branch string

	// BuildDate contains the date of the current build.
	BuildDate string

	// Version contains version
	Version string
)

func init() {
	// Log as JSON instead of the default ASCII formatter.
	log.SetFormatter(&log.TextFormatter{ForceColors: true, FullTimestamp: true, DisableTimestamp: false})

	// Output to stdout instead of the default stderr
	// Can be any io.Writer, see below for File example
	log.SetOutput(os.Stdout)

	// Only log the warning severity or above.
	log.SetLevel(log.DebugLevel)

}

func main() {

	//load app config
	conf := loadConfig()

	//setup aggregators
	buildInfo := &commons.BuildInfo{
		Version:   Version,
		Branch:    Branch,
		BuildDate: BuildDate,
	}
	twitsBuffer := info.BufferTweets(conf.ConsumerKey, conf.ConsumerSecret, conf.Token, conf.TokenSecret, conf.SearchTerm, conf.BufferSize)
	ghAggr := info.NewGitHubAggregator(conf.GitHubToken, conf.IncludeBeta)

	router := goji.NewMux()

	//404 - NOT Found middleware
	router.Use(commons.NoHandlerFound(func(w http.ResponseWriter, rq *http.Request) {
		commons.WriteJSON(http.StatusNotFound, map[string]string{"error": "not found"}, w)
	}))

	//CORS middleware, allow all domains
	router.Use(func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, rq *http.Request) {
			w.Header().Add("Access-Control-Allow-Origin", "*")
			next.ServeHTTP(w, rq)
		})

	})

	//info endpoint
	router.Handle(pat.Get("/info"), http.HandlerFunc(func(w http.ResponseWriter, rq *http.Request) {
		commons.WriteJSON(http.StatusOK, buildInfo, w)
	}))

	router.HandleFunc(pat.Get("/twitter"), func(w http.ResponseWriter, rq *http.Request) {
		if err := jsonpRS(http.StatusOK, info.GetTweets(twitsBuffer), w, rq); nil != err {
			log.Error(err)
		}
	})

	router.HandleFunc(pat.Get("/versions"), func(w http.ResponseWriter, rq *http.Request) {
		if err := jsonpRS(http.StatusOK, ghAggr.GetLatestTags(), w, rq); nil != err {
			log.Error(err)
		}
	})

	//GitHub-related routes
	ghRouter := goji.SubMux()
	router.Handle(pat.New("/github/*"), ghRouter)
	ghRouter.Handle(pat.Get("/stars"), http.HandlerFunc(func(w http.ResponseWriter, rq *http.Request) {
		commons.WriteJSON(http.StatusOK, ghAggr.GetStars(), w)
	}))
	ghRouter.Handle(pat.Get("/contribution"), http.HandlerFunc(func(w http.ResponseWriter, rq *http.Request) {
		commons.WriteJSON(http.StatusOK, ghAggr.GetContributionStats(), w)
	}))
	ghRouter.Handle(pat.Get("/issues"), http.HandlerFunc(func(w http.ResponseWriter, rq *http.Request) {
		commons.WriteJSON(http.StatusOK, ghAggr.GetIssueStats(), w)
	}))

	//aggregate everything into on rs
	router.Handle(pat.Get("/"), http.HandlerFunc(func(w http.ResponseWriter, rq *http.Request) {
		rs := map[string]interface{}{}
		rs["latest_versions"] = ghAggr.GetLatestTags()
		rs["tweets"] = info.GetTweets(twitsBuffer)
		rs["build"] = buildInfo

		ghStats := map[string]interface{}{
			"stars":              ghAggr.GetStars(),
			"contribution_stats": ghAggr.GetContributionStats(),
			"issue_stats":        ghAggr.GetIssueStats(),
		}
		rs["github"] = ghStats

		commons.WriteJSON(http.StatusOK, rs, w)
	}))

	// listen and server on mentioned port
	log.Infof("Starting on port %d", conf.Port)
	http.ListenAndServe(":"+strconv.Itoa(conf.Port), router)

}

func jsonpRS(status int, body interface{}, w http.ResponseWriter, rq *http.Request) error {
	jsonp := rq.URL.Query()["jsonp"]
	if nil != jsonp && len(jsonp) >= 1 {
		return commons.WriteJSONP(status, body, jsonp[0], w)
	}
	return commons.WriteJSON(status, body, w)
}

func loadConfig() *config {
	cfg := config{}
	err := env.Parse(&cfg)
	if err != nil {
		log.Fatalf("%+v\n", err)
	}
	return &cfg
}

type config struct {
	Port           int    `env:"PORT" envDefault:"8080"`
	ConsumerKey    string `env:"TWITTER_CONSUMER,required"`
	ConsumerSecret string `env:"TWITTER_CONSUMER_SECRET,required"`
	Token          string `env:"TWITTER_TOKEN,required"`
	TokenSecret    string `env:"TWITTER_TOKEN_SECRET,required"`
	BufferSize     int    `env:"TWITTER_BUFFER_SIZE" envDefault:"10"`
	SearchTerm     string `env:"TWITTER_SEARCH_TERM" envDefault:"@reportportal_io"`
	IncludeBeta    bool   `env:"GITHUB_INCLUDE_BETA" envDefault:"false"`
	GitHubToken    string `env:"GITHUB_TOKEN" envDefault:"false"`
}
